option cc_generic_services = true;


message BeginArgs {

};

message BeginReply {
  required int32 err = 1;
  required int32 tid = 2;
};

message AbortArgs {
  required int32 tid = 1;
};

message AbortReply {
  required int32 err = 1;
};

message EndArgs {
  required int32 tid = 1;
};

message EndReply {
  required int32 err = 1;
};

message ReadArgs {
  required int32 tid = 1;
  required string key = 2;
};

message ReadReply {
  required int32 err = 1;
  required string value = 2;
};

message WriteArgs {
  required int32 tid = 1;
  required string key = 2;
  required string value = 3;

  optional int64 clientID = 4;
  optional int64 seqID = 5;
};

message WriteReply {
  required int32 err = 1;
}


service TransactionCoordinatorBase {

  rpc BEGIN(BeginArgs) returns (BeginReply);
  rpc ABORT(AbortArgs) returns (AbortReply);
  rpc END(EndArgs) returns (EndReply);

  rpc READ(ReadArgs) returns (ReadReply);
  rpc WRITE(WriteArgs) returns (WriteReply);
};






